src/main.cpp

```diff
@@ -4439,6 +4527,28 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             pfrom->PushMessage("getheaders", chainActive.GetLocator(pindexLast), uint256());
         }
 
+        CNodeState *nodestate = State(pfrom->GetId());
+        // If this set of headers is valid and ends in a block with more work
```

Comment:
Can't this result in duplicate in-flights? The fact that it's not yet our tip does not imply we're not downloading it yet.